smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/WaiterGenerator.java

```diff
@@ -0,0 +1,223 @@
+
+            writer.openBlock("const exitConditions = [", "];", () -> {
+                writer.write("waiterTimeout(params.maxWaitTime), ");
+                writer.write("runPolling<$L, $L> (params, params.client, input, checkState)",
+                        serviceSymbol.getName(), inputSymbol.getName());
+            });
+
+            writer.openBlock("if (params.abortController) {", "};", () -> {
+                writer.write("exitConditions.push(abortTimeout(params.abortController.signal));");
+            });
+
+            writer.write("return Promise.race(exitConditions);");
+        });
+    }
+
+    private void generateImports() {
+        // service imports
+        writer.addImport(inputSymbol.getName(),
+                inputSymbol.getName(),
+                inputSymbol.getNamespace());
+        writer.addImport(operationSymbol.getName(),
+                operationSymbol.getName(),
+                operationSymbol.getNamespace());
+        writer.addImport(serviceSymbol.getName(), serviceSymbol.getName(), serviceSymbol.getNamespace());
+
+        // util waiter imports
+        writer.addImport("validateWaiterOptions", "validateWaiterOptions", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("WaiterResult", "WaiterResult", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("WaiterState", "WaiterState", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("WaiterOptions", "WaiterOptions", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("waiterTimeout", "waiterTimeout", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("runPolling", "runPolling", WAITABLE_UTIL_PACKAGE);
+        writer.addImport("abortTimeout", "abortTimeout", WAITABLE_UTIL_PACKAGE);
+
+        String serviceWaiterType = serviceSymbol.getName() + "Waiter";
+        writer.addImport(serviceWaiterType, serviceWaiterType, "./" + WAITABLE_INTERFACE_FILE.replace(".ts", ""));
+    }
+
+    private String getWaiterResult(AcceptorState resultantState) {
+        if (resultantState == AcceptorState.SUCCESS) {
+            return  "{ state: WaiterState.SUCCESS }";
+        } else if (resultantState == AcceptorState.FAILURE) {
+            return  "{ state: WaiterState.FAILURE }";
+        }
+        throw new CodegenException("Hit an invalid acceptor state to codegen " + resultantState.toString());
+    }
+
+    private void generateSuccessMatcher(Matcher.SuccessMember member, AcceptorState state) {
+        writer.write("return $L", getWaiterResult(state));
+    }
+
+    private void generateErrorMatcher(String accessor, Matcher.ErrorTypeMember member, AcceptorState state) {
+        writer.openBlock("if ($L.name && $L.name == $S) {", "}", accessor, accessor,
+                member.getValue(), () -> {
+            writer.write("return $L", getWaiterResult(state));
+        });
+    }
+
+    private void generatePathMatcher(String accessor, PathMatcher pathMatcher, AcceptorState state) {
+        writer.openBlock("try {", "} catch (e) {}", () -> {
```

Comment:
correct - theres all sorts of errors that can be thrown, null, invalid index, etc, etc. We don't want to exit the waiter in this case so we just move on to the next acceptor check. 

I suppose we could look at the full set of errors that could possibly be thrown, but I think this would be the source of a lot of code and a lot of bugs for not too much gain.